pkg/ruler/base/ruler.go

```diff
@@ -642,6 +704,66 @@ func filterRuleGroups(userID string, ruleGroups []*rulespb.RuleGroupDesc, ring r
 	return result
 }
 
+// filterRules returns a list of rule groups, each with a single rule, ONLY for rules that are "owned" by the given ring instance.
+func filterRules(userID string, ruleGroups []*rulespb.RuleGroupDesc, ring ring.ReadRing, instanceAddr string, logger log.Logger, ringCheckErrors prometheus.Counter) []*rulespb.RuleGroupDesc {
+	// Return one rule group per rule that is owned by this ring instance.
+	// One rule group is returned per rule because Prometheus executed rule groups concurrently but rules within a group sequentially;
+	// we are sharding by rule here to explicitly *avoid* sequential execution since Loki does not need this.
+	var result = make([]*rulespb.RuleGroupDesc, 0, len(ruleGroups))
+
+	for _, g := range ruleGroups {
+		logger = log.With(logger, "user", userID, "namespace", g.Namespace, "group", g.Name)
+
+		for _, r := range g.Rules {
+			logger = log.With(logger, "rule", getRuleIdentifier(r))
+
+			owned, err := instanceOwnsRule(ring, g, r, instanceAddr)
+			if err != nil {
+				ringCheckErrors.Inc()
+				level.Error(logger).Log("msg", "failed to check if the ruler replica owns the rule", "err", err)
+				continue
+			}
+
+			if !owned {
+				level.Debug(logger).Log("msg", "rule not owned, ignoring")
+				continue
+			}
+
+			level.Debug(logger).Log("msg", "rule owned")
+			// clone the group and replace the rules
+			clone, ok := proto.Clone(g).(*rulespb.RuleGroupDesc)
+			if !ok {
+				level.Error(logger).Log("msg", "failed to filter rules", "err", "failed to clone rule group; type coercion failed")
+				continue
+			}
+
+			// Prometheus relies on group names being unique, and this assumption is very deeply baked in
+			// so we append the rule token to make each group name unique.
+			// TODO(dannyk) this is all quite hacky, and we shoud look at forking Prometheus' rule evaluation engine at some point.
+			clone.Name = AddRuleTokenToGroupName(g, r)
```

Comment:
This could use your `cloneGroupWithRule` function defined in tests since it's doing the same logic

Reply:
Good call-out, thanks!